crates/bevy_ecs/src/schedule_v3/graph_utils.rs

```diff
@@ -0,0 +1,219 @@
+use std::fmt::Debug;
+
+use bevy_utils::{
+    petgraph::{graphmap::NodeTrait, prelude::*},
+    HashMap, HashSet,
+};
+use fixedbitset::FixedBitSet;
+
+use crate::schedule_v3::set::*;
+
+/// Unique identifier for a system or system set.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub(crate) enum NodeId {
+    System(usize),
+    Set(usize),
+}
+
+impl NodeId {
+    /// Returns the internal integer value.
+    pub fn index(&self) -> usize {
+        match self {
+            NodeId::System(index) | NodeId::Set(index) => *index,
+        }
+    }
+
+    /// Returns `true` if the identified node is a system.
+    pub const fn is_system(&self) -> bool {
+        matches!(self, NodeId::System(_))
+    }
+
+    /// Returns `true` if the identified node is a system set.
+    pub const fn is_set(&self) -> bool {
+        matches!(self, NodeId::Set(_))
+    }
+}
+
+/// Specifies what kind of edge should be inserted in the dependency graph.
+#[derive(Debug, Clone, Copy, Eq, PartialEq, PartialOrd, Ord, Hash)]
+pub(crate) enum DependencyEdgeKind {
+    /// A node that should be preceded.
+    Before,
+    /// A node that should be succeeded.
+    After,
+}
+
+/// Configures ambiguity detection for a single system.
+#[derive(Clone, Debug, Default)]
+pub(crate) enum Ambiguity {
+    #[default]
+    Check,
+    /// Ignore warnings with systems in any of these system sets. May contain duplicates.
+    IgnoreWithSet(Vec<BoxedSystemSet>),
+    /// Ignore all warnings.
+    IgnoreAll,
+}
+
+#[derive(Clone)]
+pub(crate) struct GraphInfo {
+    pub(crate) sets: Vec<BoxedSystemSet>,
+    pub(crate) dependencies: Vec<(DependencyEdgeKind, BoxedSystemSet)>,
+    pub(crate) ambiguous_with: Ambiguity,
+}
+
+/// Converts 2D row-major pair of indices into a 1D array index.
+pub(crate) fn index(row: usize, col: usize, num_cols: usize) -> usize {
+    debug_assert!(col < num_cols);
+    (row * num_cols) + col
+}
+
+/// Converts a 1D array index into a 2D row-major pair of indices.
+pub(crate) fn row_col(index: usize, num_cols: usize) -> (usize, usize) {
+    (index / num_cols, index % num_cols)
+}
+
+pub(crate) struct CheckGraphResults<V> {
+    /// Boolean reachability matrix for the graph.
+    pub(crate) reachable: FixedBitSet,
+    /// Pairs of nodes that have a path connecting them.
+    pub(crate) connected: HashSet<(V, V)>,
+    /// Pairs of nodes that don't have a path connecting them.
+    pub(crate) disconnected: HashSet<(V, V)>,
+    /// Edges that are redundant because a longer path exists.
+    pub(crate) transitive_edges: Vec<(V, V)>,
+    /// Variant of the graph with no transitive edges.
+    #[allow(dead_code)]
```

Comment:
Should we keep the dead code in, or remove it for now? If we're keeping it, I'd appreciate a TODO with a link to an issue.

Reply:
OK, I added the TODO. Will make an issue about the "if needed" ordering afterwards.

I could remove the field until something needed to read it, but the checking algorithm is designed to get the graph's transitive reduction and its transitive closure at the same time, so I can't really change that.

Reply:
Great, I'm fine with that.